tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs

```diff
@@ -3096,3 +3096,159 @@ where
     });
     assert!(result.is_err(), "division by zero should panic");
 }
+
+pub(crate) fn full_propagate_test<P, T>(param: P, mut executor: T)
+where
+    P: Into<PBSParameters>,
+    T: for<'a> FunctionExecutor<&'a mut RadixCiphertext, ()>,
+{
+    let (cks, sks) = KEY_CACHE.get_from_params(param);
+
+    let cks = RadixClientKey::from((cks, NB_CTXT));
+
+    // message_modulus^vec_length
+    let modulus = cks.parameters().message_modulus().0.pow(NB_CTXT as u32) as u64;
+
+    executor.setup(&cks, sks.clone());
+
+    let block_msg_mod = cks.parameters().message_modulus().0 as u64;
+    let block_carry_mod = cks.parameters().message_modulus().0 as u64;
+    let block_total_mod = block_carry_mod * block_msg_mod;
+
+    let clear_max_value = modulus - 1;
+    for msg in 1..block_msg_mod {
+        // Here we just create a block, encrypting the max message,
+        // which means its carries are empty, and test that adding
+        // something to the first block, correctly propagates
+
+        // The first block has value block_msg_mod - 1
+        // and we will add to it a message in range [1..msg_mod-1]
+        // We still have to make sure, it won't exceed the block space
+        // (which for param_message_X_carry_X is wont)
+        if (block_msg_mod - 1) + msg >= block_total_mod {
+            continue;
+        }
+
+        let max_value = cks.encrypt(clear_max_value);
+        let rhs = cks.encrypt(msg);
+
+        let mut ct = sks.unchecked_add(&max_value, &rhs);
+        executor.execute(&mut ct);
+        let decrypted_result: u64 = cks.decrypt(&ct);
+        let expected_result = clear_max_value.wrapping_add(msg) % modulus;
+        assert_eq!(
+            decrypted_result, expected_result,
+            "Invalid full propagation result, gave ct = {clear_max_value} + {msg}, \
+            after propagation expected {expected_result}, got {decrypted_result}"
+        );
+        assert!(
+            ct.blocks
+                .iter()
+                .all(|b| b.degree.0 as u64 == block_msg_mod - 1),
+            "Invalid degree after propagation"
+        );
+    }
+
+    if block_carry_mod == block_msg_mod {
+        // This test is easier to write with this assumption
+        // which, conveniently is true for our radix type
+        //
+        // In this test, we are creating a ciphertext which is at full capacity
+        // with just enough room that allows sequential (non-parallel)
+        // propagation to work
+
+        let mut expected_result = clear_max_value;
+
+        let msg = cks.encrypt(clear_max_value);
+        let mut ct = cks.encrypt(clear_max_value);
+        while sks.is_add_possible(&ct, &msg) {
+            sks.unchecked_add_assign(&mut ct, &msg);
+            expected_result = expected_result.wrapping_add(clear_max_value) % modulus;
+        }
+        assert!(ct
+            .blocks
+            .iter()
+            .all(|b| b.degree.0 as u64 == (block_total_mod - 1) - (block_msg_mod - 1)),);
```

Comment:
`- (block_carry_mod -1)` ?

Reply:
Yes I think `(block_carry_mod -1)` is the correct thing if we want to take into account  `block_carry_mod >= block_msg_mod`, but then In that case our full_propagate won't be a real full_propagate and so this test won't be correct

Reply:
Not sure about that, need to think about it 🤔

Reply:
Ok so in the case where params are such that `block_carry_mod > block_msg_mod`, the sequential propagation is fine as long as is_op_possible do proper checks.

The parallel carry propagation would require to be adapted to support that case

Reply:
feel free to open follow up issues